src/pkg/cli/composeUp.go

```diff
@@ -195,14 +202,39 @@ func WaitAndTail(ctx context.Context, project *compose.Project, client client.Fa
 		defer cancelTimeout()
 	}
 
-	errCompleted := errors.New("deployment succeeded") // tail canceled because of deployment completion
+	var deploymentStatusCh = make(chan error, 2)
+	wg := waitForDeploymentCompleted(ctx, provider, project, deploy, cancelTail, deploymentStatusCh)
+
+	tailOptions := NewTailOptionsForDeploy(deploy, since, verbose)
+	// blocking call to tail
+	TailUp(ctx, provider, project, deploy, tailOptions)
+	wg.Wait()
+
+	close(deploymentStatusCh)
+	for errDeployment := range deploymentStatusCh {
+		var errDeploymentCompleted pkg.ErrDeploymentCompleted
+		if !errors.As(errDeployment, &errDeploymentCompleted) && !(strings.Contains(errDeployment.Error(), "EOF")) {
+			return errDeployment
+		}
+	}
+
+	return nil
+}
 
+func waitForDeploymentCompleted(ctx context.Context, provider client.Provider, project *compose.Project, deploy *defangv1.DeployResponse, cancelTail context.CancelCauseFunc, deploymentStatusCh chan error) *sync.WaitGroup {
```

Comment:
This is a weird prototype:
1. returning a `WaitGroup` kinda muddles who owns the lifecycle
2. the name `waitForDeploymentCompleted` suggests this function waits for something to complete, but it doesn't: it returns "immediately", so in the very least the name is misleading.